### ISSUE TITLE
NoClassDefFoundError and "postgres" process not terminating

### ISSUE THREAD

**Issue by t9t** [slavaz/embedded-postgresql-maven-plugin#2]
At the end of the build, there is always a `NoClassDefFoundError` for me:

```
java.lang.NoClassDefFoundError: org/apache/commons/io/FilenameUtils
	at org.apache.commons.io.FileUtils.isSymlink(FileUtils.java:2924)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1534)
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:2270)
	at de.flapdoodle.embed.process.io.file.Files.forceDelete(Files.java:119)
	at ru.yandex.qatools.embed.postgresql.PostgresProcess.sendStopToPostgresqlInstance(PostgresProcess.java:174)
	at ru.yandex.qatools.embed.postgresql.PostgresProcess.stopInternal(PostgresProcess.java:141)
	at de.flapdoodle.embed.process.runtime.AbstractProcess.stop(AbstractProcess.java:170)
	at de.flapdoodle.embed.process.runtime.AbstractProcess$JobKiller.run(AbstractProcess.java:243)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: org.apache.commons.io.FilenameUtils
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	... 9 more
```

Also, when the build fails for whatever reason (some compilation error, failure in dependency resolution), the `postgres` process keeps running even after Maven has shut down (because the "`stop`" goal is never executed).

Not too long ago, I also created a similar Maven plugin and encountered the same issues. What I found was the problem is that Maven runs plugins using a separate class loader, while Embedded Postgres shuts down the process in a shutdown hook (using `Runtime.addShutdownHook(..)`). In the thread that the shutdown hook runs in, none of the classes are available, and so it fails to run at all, not stopping the postgres process, nor cleaning up temporary files.

My work-around then was to create a separate class loader, attach it to a thread, and run the Embedded Postgres code in that thread. This caused the classes to be available in the shutdown hook as well.

Unfortunately, I'm unable to share the exact code because I don't have access to it any more, but if you like I can try to replicate the work and create a pull request. It involved enumerating the plugin dependencies, and building a class loader instance from the appropriate JARs.

**Comment by slavaz** [comment]
> but if you like I can try to replicate the work and create a pull request

Yes, i will be happy with it.

**Comment by t9t** [comment]
@slavaz I created a pull request: https://github.com/slavaz/embedded-postgresql-maven-plugin/pull/3

**Comment by slavaz** [comment]
Merged and released 1.0.5 version.

[INFO] Uploaded: https://oss.sonatype.org:443/service/local/staging/deployByRepositoryId/comgithubslavaz-1004/com/github/slavaz/embedded-postgresql-maven-plugin/1.0.5/embedded-postgresql-maven-plugin-1.0.5.pom.asc (163 B at 0.9 KB/sec)
[INFO] [INFO]  * Upload of locally staged artifacts finished.
[INFO] [INFO]  * Closing staging repository with ID "comgithubslavaz-1004".

Just await a day or two until maven mirrors updated...

Thank you for the contributing!

**Comment by t9t** [comment]
Thanks, much appreciated!
